arangod/IResearch/IResearchViewCoordinator.cpp

```diff
@@ -22,19 +22,423 @@
+}
+
+arangodb::Result createLink(
+    arangodb::LogicalCollection& collection,
+    arangodb::iresearch::IResearchViewCoordinator const& view,
+    VPackSlice link,
+    VPackBuilder& builder
+) {
+  TRI_ASSERT(builder.isEmpty());
+
+  builder.openObject();
+  if (!arangodb::iresearch::mergeSlice(builder, link)
+      || !arangodb::iresearch::IResearchLinkHelper::setType(builder)
+      || !arangodb::iresearch::IResearchLinkHelper::setView(builder, view.id())) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      std::string("failed to update link definition with the view name while updating IResearch view '")
+      + std::to_string(view.id()) + "' collection '" + collection.name() + "'"
+    );
+  }
+  builder.close();
+
+  VPackBuilder tmp;
+
+  return arangodb::methods::Indexes::ensureIndex(
+    &collection, builder.slice(), true, tmp
+  );
+}
+
+}
 
 namespace arangodb {
+
+using namespace basics;
+
 namespace iresearch {
 
 /*static*/ std::shared_ptr<LogicalView> IResearchViewCoordinator::make(
     TRI_vocbase_t& vocbase,
-    arangodb::velocypack::Slice const& info,
+    velocypack::Slice const& info,
     uint64_t planVersion,
-    LogicalView::PreCommitCallback const& preCommit
+    LogicalView::PreCommitCallback const& /*preCommit*/
+) {
+  auto view = std::unique_ptr<IResearchViewCoordinator>(
+    new IResearchViewCoordinator(vocbase, info, planVersion)
+  );
+
+  std::string error;
+
+  auto properties = info.get(PROPERTIES_FIELD);
+
+  if (!properties.isObject()) {
+    // set to defaults
+    properties = VPackSlice::emptyObjectSlice();
+  }
+
+  auto& meta = view->_meta;
+
+  if (!meta.init(properties, error)) {
+    LOG_TOPIC(WARN, iresearch::TOPIC)
+        << "failed to initialize IResearch view from definition, error: " << error;
+
+    return nullptr;
+  }
+
+  auto const links = info.get(LINKS_FIELD);
+
+  if (links.isObject()) {
+    auto& builder = view->_links;
+
+    size_t idx = 0;
+    for (auto link : velocypack::ObjectIterator(links)) {
+      auto name = link.key;
+
+      if (!name.isString()) {
+        LOG_TOPIC(WARN, iresearch::TOPIC)
+            << "failed to initialize IResearch view link from definition at index " << idx
+            << ", link name is not string";
+
+        return nullptr;
+      }
+
+      builder.add(name);
+      builder.openObject();
+      auto const res = IResearchLinkHelper::normalize(builder, link.value, false);
+
+      if (!res.ok()) {
+        LOG_TOPIC(WARN, iresearch::TOPIC)
+            << "failed to initialize IResearch view link from definition at index " << idx
+            << ", error: " << error;
+
+        return nullptr;
+      }
+      builder.close();
+    }
+  }
+
+  return view;
+}
+
+IResearchViewCoordinator::IResearchViewCoordinator(
+    TRI_vocbase_t& vocbase,
+    velocypack::Slice info,
+    uint64_t planVersion
+) : LogicalView(vocbase, info, planVersion) {
+  TRI_ASSERT(ServerState::instance()->isCoordinator());
```

Comment:
Do we not need these objects on dbservers occasionally? Representing cache values?

Reply:
I believe our concept is to segregate responsibilities, i.e. having IResearchViewCoordinator - on coordinator and IResearchViewDBServer on db servers.
Moreover it's impossible to register 2 different factories for a single view type.